Fix pick failure with rectilinear grids and external-to-problem ghosts.

biagas opened the pull request:
Don't user rectilinear grid fast-path in Locate queries if ghosts
are present. The slower path is faster than removing ghosts from
the rectilinear grid.

### Description

Resolves #17014

### Type of change

<!-- Please check one of the boxes below -->

* [X] Bug fix
* [ ] New feature
* [ ] Documentation update
* [ ] Other <!-- please explain with a note below -->

### How Has This Been Tested?

I ran the test suite with success.
Pick now succeeds with the user's test data.

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [X] I have followed the [style guidelines][1] of this project.~~
- [X] I have performed a self-review of my own code.~~
- [X] I have commented my code where applicable.~~
- [X] I have updated the release notes.~~
~~- [ ] I have made corresponding changes to the documentation.~~
~~- [ ] I have added debugging support to my changes.~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works.~~
- [X] I have confirmed new and existing unit tests pass locally with my changes.~~
~~- [ ] I have added new baselines for any new tests to the repo.~~
- [X] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~
- [X] I have assigned reviewers (see [VisIt's PR procedures][2] for more information).~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

biagas committed:
Fix pick failure with rectilinear grids and external-to-problem ghosts.
Don't user rectilinear grid fast-path in Locate queries if ghosts
are present. The slower path is faster than removing ghosts from
the rectilinear grid.

biagas committed:
Remove include leftover from testing

biagas committed:
Remove ifdef leftover from testing.

markcmiller86 commented on src/avt/Queries/Pick/avtLocateCellQuery.C:
```diff
@@ -167,8 +173,10 @@ avtLocateCellQuery::Execute(vtkDataSet *ds, const int dom)
+    // Don't use the RectilinearGrid fast-path if there are ghosts, as it
+    // takes longer to remove the ghosts than to use the fast-path.
+    if ( ds->GetDataObjectType() != VTK_RECTILINEAR_GRID ||
+         dataAtts.GetContainsGhostZones())
```
So, just to validate the logic here...any mesh block that is either NOT rectilinear or HAS ghost zones will wind up going through the slower logic to find the picked cell instead of using the potentially faster `RGridFindCell()` logic, right?

In practice, in any parallel decomposed mesh, even one with all rectilinear blocks, every block is likely going to have *some* ghost zones (well, maybe just ghost *nodes* is the common case where neighboring blocks abut) and so almost exclusively now, we'll always pick the slower path.

I know your modification comment in the preamble mentions something about performance. Do you have a good feel for impact on our large multi-block, structured gridded codes? Maybe those aren't typically truly rectilinear anyways.

biagas replied:
I am not sure about the impact for real-sized data. I just did some timing comparisons with our test data.
I decided to look into changing the ghost test to GetContainsExternalBoundaryGhosts (which satisfies the users dataset).  Am currently evaluating the regression result differences for said change.

biagas replied:
@markcmiller86 Changing the test to GetContainsExternalBoundaryGhosts reduces the amount of times pick will not take the rectilinear grid fast-path. (at least for the pick tests in our repo).   I think this should be sufficient, what do you think?

markcmiller86 replied:
Sounds good. I say we ship it :shipit:

markcmiller86 reviewed (dismissed):
Just wanted to confirm your thoughts on one point...

biagas committed:
Change ghost for rectilinear grids to ContainsExteriorBoundaryGhosts

Reduces how frequently the locate query will NOT take the rectilinear grid fast-path.

biagas committed:
Update commit comment.

biagas committed:
Fix commit comment.
